### PR TITLE
570 external git server

### LINKED ISSUE #570: Support for using external Gitlab instead of gitea on zarf init
**Is your feature request related to a problem? Please describe.**
Similar to #560, it may be desirable for some to use an external Gitlab rather than the internal gitea instance that is shipped with Zarf.

**Describe the solution you'd like**
In my case specifically my platform has a pre-existing instance of Gitlab that I would like the zarf sync'd code to reside in.

**Describe alternatives you've considered**
Could configure Gitlab to mirror the gitea instance but I couldn't figure a way to do this in an automated fashion.  This work-around still doesn't prevent data duplication.

### PR DESCRIPTION
**Pull request by YrrepNoj**
Fixes #570

### CODE REVIEW

File: packages/zarf-agent/zarf.yaml [hunk 1]
```diff
@@ -7,9 +7,9 @@ components:
   - name: zarf-agent
     required: true
     images:
-      - "defenseunicorns/zarf-agent:992efd"
+      - "defenseunicorns/zarf-agent:remote-git-wip-agent"
     variables:
-      tag: "992efd"
+      tag: "remote-git-wip-agent"    # TODO: Replace with a new real build
```
Comment (YrrepNoj) [inline packages/zarf-agent/zarf.yaml:12]: Don't forget to change this

File: Makefile [hunk 2]
```diff
@@ -70,13 +70,13 @@ dev-agent-image:
...
 	CGO_ENABLED=0 GOOS=linux go build -o build/zarf-linux-$(arch) main.go
 	DOCKER_BUILDKIT=1 docker build --tag $(tag) --build-arg TARGETARCH=$(arch) . && \
-	kind load docker-image zarf-agent:$(tag) && \
+	kind load docker-image $(tag) && \
 	kubectl -n zarf set image deployment/agent-hook server=$(tag)
 
 init-package: ## Create the zarf init package, macos "brew install coreutils" first
 	@test -s $(ZARF_BIN) || $(MAKE) build-cli
 
-	@test -s ./build/zarf-init-$(ARCH).tar.zst || $(ZARF_BIN) package create -o build -a $(ARCH) --confirm .
```
Comment (jeff-mccoy) [inline Makefile:79]: I dont hate this change, except this will force local tests to always rebuild the init package--which often isn't necessary during dev, is there a particular reason to keep the behavior consistent with the examples?

File: packages/zarf-agent/manifests/deployment.yaml [hunk 3]
```diff
@@ -41,7 +41,13 @@ spec:
             - name: tls-certs
               mountPath: /etc/certs
               readOnly: true
+            - name: zarf-state
```
Comment (jeff-mccoy) [inline packages/zarf-agent/manifests/deployment.yaml:44]: This shouldn't have been as exciting as it was for me to see.....

File: src/internal/git/push.go [hunk 4]
```diff
@@ -18,12 +18,38 @@ const offlineRemoteName = "offline-downstream"
...
+		gitServerURL = gitServerInfo.GitAddress
+
+		// Add a port to the URL if it was provided
+		if gitServerInfo.GitPort != 0 {
+			gitServerURL = fmt.Sprintf("%s:%d", gitServerInfo.GitAddress, gitServerInfo.GitPort)
+		}
 
-	tunnelUrl := fmt.Sprintf("http://%s", tunnel.Endpoint())
+		// If this URL points to a resource within the cluster, create a tunnel to it
+		if strings.Contains(gitServerURL, "svc.cluster.local") {
```
Comment (Racer159) [inline src/internal/git/push.go:42]: This is probably fine since this is super edge-casey, but my brain immediately goes to "but what about https://svc.cluster.local.example.com!"

**Review by Racer159 — APPROVED** [review]

File: packages/zarf-agent/zarf.yaml (hunk 1 above)
Comment (Racer159) [reply packages/zarf-agent/zarf.yaml:12]: Approved assuming this is changed

File: Makefile (hunk 2 above)
Comment (YrrepNoj) [reply Makefile:79]: `make e2e` builds the example packages as a side affect.
`make init-package` builds the init-package as its primary purpose. 

When I was testing changes with the zarf-agent it took me longer than I'm proud to admit to realize that I wasn't actually rebuilding the init-package 😛

File: src/internal/git/push.go (hunk 4 above)
Comment (YrrepNoj) [reply src/internal/git/push.go]: That's a good edge-case! I changed the way we check the URL string to avoid situations like this.

Comment (Racer159) [reply src/internal/git/push.go]: :tada:

**Review by Racer159 — APPROVED** [review]

File: src/test/external-git/secret.yaml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+apiVersion: v1
+kind: Secret
+metadata:
+  name: private-git-server
+  namespace: git-server
+type: Opaque
+data:
+  username: Z2l0LXVzZXI=
+  password: c3VwZXJTZWN1cmVQYXNzd29yZA==
```
Comment (jeff-mccoy) [inline src/test/external-git/secret.yaml:9]: Some scanners will 🤮 on this username/password.  Clearly this is just a test entry, but could we leave come comment stating that in case someone's scanner trips up on it?

Comment (YrrepNoj) [reply src/test/external-git/secret.yaml:9]: good point, will do 👍

File: Makefile [hunk 6]
```diff
@@ -79,13 +79,12 @@ dev-agent-image:
...
 	DOCKER_BUILDKIT=1 docker build --tag $(tag) --build-arg TARGETARCH=$(arch) . && \
-	kind load docker-image zarf-agent:$(tag) && \
+	kind load docker-image $(tag) && \
 	kubectl -n zarf set image deployment/agent-hook server=$(tag)
 
 init-package: ## Create the zarf init package, macos "brew install coreutils" first
 	@test -s $(ZARF_BIN) || $(MAKE) build-cli
-
-	@test -s ./build/zarf-init-$(ARCH).tar.zst || $(ZARF_BIN) package create -o build -a $(ARCH) --set AGENT_IMAGE=$(AGENT_IMAGE) --confirm .
+	$(ZARF_BIN) package create -o build -a $(ARCH) --set AGENT_IMAGE=$(AGENT_IMAGE) --confirm .
```
Comment (jeff-mccoy) [inline Makefile:87]: This still kinda kills me since in dev we're now forcing an init package rebuilding on every test run--this will slow down my flow a good bit, can you talk about how you see this flow working if you run e2e tests locally often?

File: src/cmd/initialize.go [hunk 7]
```diff
@@ -103,7 +103,13 @@ func init() {
...
 	initCmd.Flags().StringVar(&config.DeployOptions.Components, "components", "", "Comma-separated list of components to install.")
-	initCmd.Flags().StringVar(&config.DeployOptions.StorageClass, "storage-class", "", "Describe the StorageClass to be used")
-	initCmd.Flags().StringVar(&config.DeployOptions.Secret, "secret", "", "Root secret value that is used to 'seed' other secrets")
-	initCmd.Flags().StringVar(&config.DeployOptions.NodePort, "nodeport", "", "Nodeport to access the Zarf container registry. Between [30000-32767]")
+	initCmd.Flags().StringVar(&config.InitOptions.StorageClass, "storage-class", "", "Describe the StorageClass to be used")
+	initCmd.Flags().StringVar(&config.InitOptions.Secret, "secret", "", "Root secret value that is used to 'seed' other secrets")
+	initCmd.Flags().StringVar(&config.InitOptions.NodePort, "nodeport", "", "Nodeport to access the Zarf container registry. Between [30000-32767]")
+
+	// Flags for using an external Git server
+	initCmd.Flags().StringVar(&config.InitOptions.GitServerInfo.GitAddress, "git-url", "", "External git server url to use for this Zarf cluster")
```
Comment (jeff-mccoy) [inline src/cmd/initialize.go:111]: Not a huge deal, but `config.InitOptions.GitServerInfo.Gitaddress` feels fairly verbose to me.  `config.InitOptions.GitServer.Address` I think captures the same info.

File: src/cmd/initialize.go [hunk 8]
```diff
@@ -103,7 +103,13 @@ func init() {
...
+	initCmd.Flags().StringVar(&config.InitOptions.GitServerInfo.GitPushUsername, "git-user", "", "Username to connect to the external git server. User must be able to create repositories via 'git push'")
```
Comment (jeff-mccoy) [inline src/cmd/initialize.go:112]: Is there a way in the flags to set the read-only username?  I imagine someone using an external registry would actually be more likely to also have different accounts or tokens to access rw vs r.

File: src/cmd/tools.go [hunk 9]
```diff
@@ -73,7 +73,7 @@ var readCredsCmd = &cobra.Command{
 		// Continue loading state data if it is valid
 		config.InitState(state)
 
-		fmt.Println(config.GetSecret(config.StateGitPush))
+		fmt.Println(state.GitServerInfo.GitPushPassword)
```
Comment (jeff-mccoy) [inline src/cmd/tools.go:76]: Doesnt have to be in this PR overall, but would be nice to state (maybe via`message` so it's on stderr) that this is only the Git Server Push Password, not every credential.

File: src/internal/agent/hooks/flux.go [hunk 10]
```diff
@@ -34,15 +36,34 @@ func NewGitRepositoryMutationHook() operations.Hook {
 func mutateGitRepository(r *v1.AdmissionRequest) (*operations.Result, error) {
 	var patches []operations.PatchOperation
 
+	zarfState, err := getZarfStateFromFileWithinAgentPod(zarfStatePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load zarf state from file: %v", err)
+	}
+
+	// Default to the InCluster gitURL
+	gitServerURL := config.ZarfInClusterGitServiceURL
```
Comment (jeff-mccoy) [inline src/internal/agent/hooks/flux.go:45]: This seems strange to me, why wouldn't we just parse the state for the URL like we do for external?  I'd personally rather keep it generic and since this is breaking change anyway, feels like the right time to move to the new pattern.  Is there any benefit to tracking in-cluster git server otherwise?

File: src/internal/git/push.go [hunk 11]
```diff
@@ -18,12 +18,38 @@ const offlineRemoteName = "offline-downstream"
...
 
 func PushAllDirectories(localPath string) error {
-	// Establish a git tunnel to send the repos
-	tunnel := k8s.NewZarfTunnel()
-	tunnel.Connect(k8s.ZarfGit, false)
-	defer tunnel.Close()
+	gitServerInfo := config.GetGitServerInfo()
+	gitServerURL := ""
+
+	if gitServerInfo.InternalServer {
```
Comment (jeff-mccoy) [inline src/internal/git/push.go:24]: It feels like it would be cleaner to just use the gitServerInfo data to build a tunnel if it's in cluster (by detecting if the url ends with `svc.cluster.local`).

File: src/internal/git/push.go [hunk 12]
```diff
@@ -37,55 +63,68 @@ func PushAllDirectories(localPath string) error {
...
 
 		// Add the read-only user to this repo
-		repoPathSplit := strings.Split(path, "/")
-		repoNameWithGitTag := repoPathSplit[len(repoPathSplit)-1]
-		repoName := strings.Split(repoNameWithGitTag, ".git")[0]
-		err = addReadOnlyUserToRepo(tunnelUrl, repoName)
-		if err != nil {
-			message.Warnf("Unable to add the read-only user to the repo: %s\n", repoName)
-			return err
+		if gitServerInfo.InternalServer {
```
Comment (jeff-mccoy) [inline src/internal/git/push.go:79]: We should probably think about this some more--we intentionally kept the read only user creation in the git server component so it wasn't tightly-coupled/hard-coded.  

It may require a different PR, but I think I've found a better way that exists now in Gitea to handle read only users:  you can now create repos per-org on push and then assign a user to the org once--or just set the rules to "must be logged in" to gitea and even that isn't required.    https://docs.gitea.io/en-us/config-cheat-sheet/#repository-repository, `ENABLE_PUSH_CREATE_ORG`. 


Screen Shot 2022-08-01 at 1 42 19 PM
Screen Shot 2022-08-01 at 1 42 11 PM
Screen Shot 2022-08-01 at 1 41 34 PM
Screen Shot 2022-08-01 at 1 40 53 PM
Screen Shot 2022-08-01 at 1 39 55 PM
Screen Shot 2022-08-01 at 1 38 55 PM

File: src/internal/k8s/tunnel.go [hunk 13]
```diff
@@ -88,6 +88,33 @@ func PrintConnectTable() error {
 	return nil
 }
 
+// NewTunnelFromServiceURL takes a serviceURL and parses it to create a tunnel to the cluster. The string is expected to follow the following format:
```
Comment (jeff-mccoy) [inline src/internal/k8s/tunnel.go:91]: I love this feature

File: src/internal/k8s/tunnel.go [hunk 14]
```diff
@@ -88,6 +88,33 @@ func PrintConnectTable() error {
...
+// Example serviceURL: http://{SERVICE_NAME}.{NAMESPACE}.svc.cluster.local:{PORT}
+func NewTunnelFromServiceURL(serviceURL string) (*Tunnel, error) {
+
+	// Remove the protocol from the URL
+	serviceURL = strings.TrimPrefix(serviceURL, "http://")
```
Comment (jeff-mccoy) [inline src/internal/k8s/tunnel.go:96]: Think we'd be better served using a URL parsing library than manually hacking this:  https://golangbyexample.com/parse-url-golang/

File: Makefile (hunk 2 above)
Comment (jeff-mccoy) [reply Makefile]: Yeah that's fair, then can we make a new `make init-package-if-not-exist` or something that e2e calls--and we could do the same for examples too I suppose, but that will make things noisy in the makefile.  Alternatively, we could just dump a warning in the makefile output, that the file already exists so reusing it, this is just for devs after all.  You can change text colors in make pretty easy:

```shell
GREEN  := $(shell tput -Txterm setaf 2)
YELLOW := $(shell tput -Txterm setaf 3)
WHITE  := $(shell tput -Txterm setaf 7)
CYAN   := $(shell tput -Txterm setaf 6)
RESET  := $(shell tput -Txterm sgr0)
```

```shell
 ${YELLOW}words in yellow${RESET} ${GREEN}words in green${RESET}
```

**Review by jeff-mccoy — COMMENTED** [review]
Will continue PR review after we work through these comments--this is going to be a great option for larger deployments and lays the needed groundwork for external registry support 🥳

File: src/internal/agent/hooks/flux.go (hunk 10 above)
Comment (Racer159) [reply src/internal/agent/hooks/flux.go:45]: FWIW I had the same thought reviewing this earlier, I just didn't write it down.

File: src/cmd/initialize.go (hunk 8 above)
Comment (YrrepNoj) [reply src/cmd/initialize.go:112]: I left that out on purpose for now. 

We use the read-only user/pass when we create the secrets within the `post-render.go` logic. If we add flags for read-only user account(s) those credentials will not work because we directly use the Gitea API and we won't be able to do that with external registries.

File: Makefile (hunk 2 above)
Comment (YrrepNoj) [reply Makefile]: How about a middle ground where we keep the `init-package` explicit and always creates the package but then I update the workflows.yaml and the `test-e2e` target to only make the init-package if it's not already there?

File: src/cmd/tools.go (hunk 9 above)
Comment (YrrepNoj) [reply src/cmd/tools.go:76]: In the same vein, changing the command name might be useful `get-admin-password` is pretty generic and unclear unless your read the description in a help menu.

File: src/internal/agent/hooks/flux.go (hunk 10 above)
Comment (YrrepNoj) [reply src/internal/agent/hooks/flux.go:45]: I'll make this part more generic.

The only other time where it makes sense to track in-cluster git server is when we are creating a read-only user. Since we use Gitea api we can't do that for an external server.

File: Makefile (hunk 2 above)
Comment (jeff-mccoy) [reply Makefile]: that seems reasonable to me

File: Makefile [hunk 15]
```diff
@@ -79,19 +79,20 @@ dev-agent-image:
...
 
 ci-release: init-package ## Create the init package
 
 build-examples:
 	@test -s $(ZARF_BIN) || $(MAKE) build-cli
 
+	@test -s ./build/zarf-init-$(ARCH).tar.zst || $(ZARF_BIN) package create -o build -a $(ARCH) --set AGENT_IMAGE=$(AGENT_IMAGE) --confirm .
```
Comment (jeff-mccoy) [inline Makefile:94]: I don't really see `build-examples` depending on `init-package`, think this test probably makes more sense under the e2e section.

File: .github/workflows/test-k3d.yml [hunk 16]
```diff
@@ -49,7 +49,7 @@ jobs:
           docker buildx build --push --platform linux/amd64 --tag zarfdev/agent:$GITHUB_SHA .
 
       - name: "Make Packages"
-        run: make init-package build-examples ARCH=amd64 AGENT_IMAGE=zarfdev/agent:$GITHUB_SHA
+        run: make build-examples ARCH=amd64 AGENT_IMAGE=zarfdev/agent:$GITHUB_SHA
```
Comment (jeff-mccoy) [inline .github/workflows/test-k3d.yml:52]: Probably cleaner to keep this explicit as it's going to be created anyway since it won't exist in CI

File: src/cmd/prepare.go [hunk 17]
```diff
@@ -37,7 +37,7 @@ var prepareTransformGitLinks = &cobra.Command{
 
 		// Perform git url transformation via regex
 		text := string(content)
-		processedText := git.MutateGitUrlsInText(host, text)
+		processedText := git.MutateGitUrlsInText(host, text, config.InitOptions.GitServer.PushUsername)
```
Comment (jeff-mccoy) [inline src/cmd/prepare.go:40]: This one is a little funny to me as the username is now dynamic and a `package deploy` configuration, not `package create` configuration.  Should this default to the internal user, but have an option to specify a custom one?  

Also curious how this works with external git servers--we can predict the path based off the username, is that always true in other git servers or is that something we need to talk about too?

Comment (jeff-mccoy) [reply src/cmd/prepare.go:40]: ... and scrolling down I see the flag for the pus user option, so disregard the first point.

File: src/internal/packager/seed.go [hunk 18]
```diff
@@ -103,14 +103,32 @@ func preSeedRegistry(tempPath tempPaths) {
 	}
 
 	// CLI provided overrides that haven't been processed already
-	if config.DeployOptions.NodePort != "" {
-		state.NodePort = config.DeployOptions.NodePort
+	if config.InitOptions.NodePort != "" {
```
Comment (jeff-mccoy) [inline src/internal/packager/seed.go:106]: Moving to InitOptions now made a lot of sense and will keep things saner. ✅

File: src/internal/packager/seed.go [hunk 19]
```diff
@@ -103,14 +103,32 @@ func preSeedRegistry(tempPath tempPaths) {
...
+	if config.InitOptions.StorageClass != "" {
+		state.StorageClass = config.InitOptions.StorageClass
+	}
+
+	if config.InitOptions.GitServer.Address != "" {
+		// The user has provided external git server info, just use what they gave!
+		state.GitServer = config.InitOptions.GitServer
+
+		// For external clusters, the read-user is the same as the push-user
+		state.GitServer.ReadUsername = state.GitServer.PushUsername
```
Comment (jeff-mccoy) [inline src/internal/packager/seed.go:121]: I still think we need to look at this--I don't see why wouldn't allow the user to specify a read-only user as well--it would likely be a robot account, PAT or similar to grant less-privilege and I think it would be a mistake to not allow them to do that.  Think this is related to assuming we can push to a different git server just like we do for gitea.

File: src/internal/packager/seed.go [hunk 20]
```diff
@@ -103,14 +103,32 @@ func preSeedRegistry(tempPath tempPaths) {
...
+		state.GitServer.ReadPassword = state.GitServer.PushPassword
+	} else {
+		// Set the GitServerInfo for the internal Gitea server
+		state.GitServer.Address = config.ZarfInClusterGitServiceURL
+		state.GitServer.Port = config.ZarfInClusterGitServicePort
+		state.GitServer.PushUsername = config.ZarfGitPushUser
+		state.GitServer.PushPassword = utils.RandomString(24)
```
Comment (jeff-mccoy) [inline src/internal/packager/seed.go:128]: this is a welcome change....

File: src/cmd/initialize.go (hunk 8 above)
Comment (jeff-mccoy) [reply src/cmd/initialize.go]: I still don't think that's really an issue, we just have to communicate "if you specify a read-only user, you must ensure that repos zarf pushes can be accessed by that user"

**Review by jeff-mccoy — COMMENTED** [review]
Left more comments, happy to discuss further--think it's getting pretty close and is a major enhancement to zarf's deployment flexibility, thanks!

File: src/internal/packager/seed.go (hunk 19 above)
Comment (YrrepNoj) [reply src/internal/packager/seed.go:121]: Let me know when you have a bit of time to chat about this. Just want to make sure we're both on the same page :)

File: src/cmd/prepare.go (hunk 17 above)
Comment (YrrepNoj) [reply src/cmd/prepare.go:40]: It seems to be fairly consistent with all of the git servers I have used before. The path is the domain followed by either the username or org name. 

I reworded the flag message here to make it a little more obvious what to do if the repo is originally pushed under an org.

Comment (Racer159) [reply src/cmd/prepare.go:40]: That is how usernames work in GitLab, GitHub, BitBucket, and GitTea (and any other host based on those like CodeBerg) so it is likely good enough (and even if they do something like `https://example.com/scm/username/project.git` instead of a subdomain you should be able to use the full `https://example.com/scm/` as the `host` variable).

File: src/internal/agent/hooks/flux.go [hunk 21]
```diff
@@ -34,15 +36,28 @@ func NewGitRepositoryMutationHook() operations.Hook {
 func mutateGitRepository(r *v1.AdmissionRequest) (*operations.Result, error) {
 	var patches []operations.PatchOperation
 
+	zarfState, err := getZarfStateFromFileWithinAgentPod(zarfStatePath)
```
Comment (Racer159) [inline src/internal/agent/hooks/flux.go:39]: Could this be shortened to just `getZarfStateFromAgent` (or similar)? `getZarfStateFromFileWithinAgentPod` is a little long

File: src/internal/agent/hooks/flux.go [hunk 22]
```diff
@@ -34,15 +36,28 @@ func NewGitRepositoryMutationHook() operations.Hook {
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to load zarf state from file: %v", err)
+	}
+
+	// Form the gitServerURL from the state
+	gitServerURL := zarfState.GitServer.Address
+	if zarfState.GitServer.Port != 0 {
+		gitServerURL += fmt.Sprintf(":%d", zarfState.GitServer.Port)
```
Comment (Racer159) [inline src/internal/agent/hooks/flux.go:47]: What I said above is not true though when using a non-default port.  If I hosted gittea behind an nginx server and rewrote a subpath to point to my git server then this would result in something like: https://example.com/my-scm:5000...

File: src/internal/git/push.go [hunk 23]
```diff
@@ -18,12 +18,25 @@ const offlineRemoteName = "offline-downstream"
...
-	tunnel := k8s.NewZarfTunnel()
-	tunnel.Connect(k8s.ZarfGit, false)
-	defer tunnel.Close()
+	gitServerInfo := config.GetGitServerInfo()
+	gitServerURL := gitServerInfo.Address
 
-	tunnelUrl := fmt.Sprintf("http://%s", tunnel.Endpoint())
+	// Add a port to the URL if it was provided
+	if gitServerInfo.Port != 0 {
+		gitServerURL = fmt.Sprintf("%s:%d", gitServerInfo.Address, gitServerInfo.Port)
```
Comment (Racer159) [inline src/internal/git/push.go:26]: Same thing here if address is not a root path

File: src/test/external-git/README.md [hunk 24]
```diff
@@ -0,0 +1,33 @@
...
+> Note: In this case, the 'external' Git server is a server that already existed inside the k8s cluster
+
+This directory holds the tests that verify Zarf can initialize a cluster to use an already existing Git server that is external to the resources Zarf manages. The tests in this directory are currently only run when manually executed.
+
+
+## Running Tests Locally
+
+### Dependencies
+Running the tests locally have the same prerequisites as running and building Zarf:
+ 1. GoLang >= `1.18.x`
```
Comment (Racer159) [inline src/test/external-git/README.md:11]: 1.19.x

**Review by Racer159 — CHANGES_REQUESTED** [review]

File: src/internal/agent/hooks/flux.go (hunk 22 above)
Comment (YrrepNoj) [reply src/internal/agent/hooks/flux.go:47]: In that case you could add the port as a string within the server URL and just not populate the port flag. If we ever actually need the port we could try to parse the port using `net/url`.

@Racer159 Do you think this works or do you think we need to come up with a more robust solution?

Comment (Racer159) [reply src/internal/agent/hooks/flux.go:47]: At that point why specify the port as a separate thing at all?

Comment (YrrepNoj) [reply src/internal/agent/hooks/flux.go:47]: That's a good point, I thought it would have been needed earlier but I'll just give the url

**Comment by YrrepNoj** [comment]
Closing this as it's being duplicated by #666 along with other features (handling an external container registry)
